### PR TITLE
[codecs] Extending codecs definitions for better developer comfort

### PR DESCRIPTION
**Pull request by martinheidegger**
This PR updates the `codecs` type definitions with new util types that makes it easier to use `codecs` with other libraries.

When working with the recently added `codecs` definitions I noticed that they are kind of stressful to work that this solves:

1. `NamedCodec` always required a type so I had to do like `NamedCodec<string>`. Now it does have a default and I can just reference `NamedCodec` without setting the default.
2. When having a structure like `hypercore` that [accepts a codec as parameter](https://github.com/hypercore-protocol/hypercore/blob/0534b7a2b2b509880773814e7e72a8a01fd4c92f/index.js#L1294) then the `hypercore` instance is of the type of the input character. `HyperCode<InputType>` but since `codecs` has fallbacks, its rather tricky to evaluate this. This PR adds a `InType` and `OutType` helper that allows to easily derive what the output of a codec is:
    ```typescript
    import { OutType } = require('codecs')
    class Hypercore<ValueCodec> {
       constructor (options: HypercoreOptions<ValueCodec>)
       get (item: number, cb: Callback<OutType<ValueCodec>>): void
    }
    ```
3. Additionally this PR adds a `CodecName` helper that allows to get the name of a selected codec.
    ```typescript
    import { CodecName } = require('codecs')
    class ToBeSerialized<ValueCodec> {
       constructor (options: HypercoreOptions<ValueCodec>)
       get name (): CodecName<ValueCodec>
    }
    ```
4. When my code had a different default codec it was kind of annoying to see `NamedCodec<'json', any, JsonValue>` so this PR also named all the given codecs: `JsonCodec`, `AsciiCodec`, `Base64Codec`, ...; In combination with the new `MaybeCodecInput` this allows following update:
    ```typescript
    // old
    createThing <T extends BaseCodec | string  | null  | undefined>(codec: T): Codec<T, NamedCodec<'json', any, JsonValue>>
    // new
    createThing <T extends CodecInput>(codec: T): Codec<T, JsonCodec>
    ```
5. Unfortunately the current code missed a definition allowing an `undefined` value to be passed in to codecs:

     ```typescript
     codecs() // worked!
     codecs(undefined) // didn't work
     ```

     This PR also fixes that small issue.

     
Unfortunately there was some update `ExpectType` with Typescript 3.7 that resolves to different types (`JsonCodec` is `JsonCodec` from 3.7 but `NamedCodec<"json", any, JsonValue>` before that). In order to get the tests to pass I had to duplicate the whole definitions and add a `3.6` variant.

### CODE REVIEW

File: types/codecs/index.d.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /// <reference types="node" />
+type ArgsType<T> = T extends (...args: infer U) => any ? U : never;
```
Comment (Tangent128) [inline types/codecs/index.d.ts:7]: This utility type looks unused by anything in the package.

File: types/codecs/ts3.6/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/// <reference types="node" />
+type ArgsType<T> = T extends (...args: infer U) => any ? U : never;
```
Comment (Tangent128) [inline types/codecs/ts3.6/index.d.ts:2]: This utility type looks unused by anything in the package.

**Review by Tangent128 — APPROVED** [review]

**Comment by martinheidegger** [comment]
Ready to merge

**Review by Tangent128 — APPROVED** [review]

**Comment by martinheidegger** [comment]
Ready to merge

**Pull request merged**